app/models/personal_access_token.rb

```diff
@@ -19,23 +20,48 @@ class PersonalAccessToken < ApplicationRecord
   attr_accessor :token_value
 
   def self.authenticate_user(user, token)
-    token = self.active.find_by(user: user, token: hash_token(user, token))
+    token = self.active.find_by(user: user, token: hash_token(user, token, :bcrypt)) ||
+      self.active.find_by(user: user, token: hash_token(user, token, :sha1))
     return false unless token
+
     token.update(last_used_at: Time.current.utc)
     true
   end
 
-  def self.token_salt(user)
-    Digest::SHA1.hexdigest(user.id.to_s)
+  def self.token_salt_bcrypt(user_id)
+    "$2a$#{Setting[:bcrypt_cost].to_s.rjust(2, '0')}$#{Digest::SHA1.hexdigest(user_id.to_s)}"
+  end
+
+  # static salt based on user id because input is already good enough (sha1 string)
+  def self.token_salt(user, method = :bcrypt)
+    if method == :bcrypt
+      token_salt_bcrypt(user.id)
+    elsif method == :sha1
+      Digest::SHA1.hexdigest(user.id.to_s)
+    else
+      raise(Foreman::Exception.new(N_("Unknown password salt method: %s"), method))
+    end
   end
 
-  def self.hash_token(user, token)
-    Digest::SHA1.hexdigest([token, token_salt(user)].join)
+  def self.hash_token(user, token, method = :bcrypt)
+    if method == :bcrypt
```

Comment:
I believe this would look a lot cleaner of we'd extract this to two classes.

```ruby
def self.hash_token(user, token, method = :bcrypt)
  hash_implementation(method).hash_token(user, token)
end

def self.hash_implementation(method = :bcrypt)
   method == :sha ? ShaHash : BcryptHash
end

class ShaHash
   def hash_token(payload, token)
      ...
   end

   def token_salt(payload)
     ...
   def
end
```

We could then reuse this for the `User` model and test it properly.

Reply:
Good idea, implemented whole new class.